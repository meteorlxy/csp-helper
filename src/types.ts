/**
 * Content Security Policy record type.
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy
 */
export interface ContentSecurityPolicyRecord {
  /* Fetch directives */
  'child-src': string;
  'connect-src': string;
  'default-src': string;
  'fenced-frame-src': string;
  'font-src': string;
  'frame-src': string;
  'img-src': string;
  'manifest-src': string;
  'media-src': string;
  'object-src': string;
  'script-src': string;
  'script-src-elem': string;
  'script-src-attr': string;
  'style-src': string;
  'style-src-elem': string;
  'style-src-attr': string;
  'worker-src': string;

  /* Document directives */
  'base-uri': string;
  'sandbox': string;

  /* Navigation directives */
  'form-action': string;
  'frame-ancestors': string;

  /* Reporting directives */
  'report-to': string;

  /* Other directives */
  'require-trusted-types-for ': string;
  'trusted-types': string;
  'upgrade-insecure-requests': '';

  /* Deprecated directives */
  /** @deprecated */
  'block-all-mixed-content': '';
  /** @deprecated */
  'prefetch-src': string;
  /** @deprecated */
  'report-uri': string;
}

/**
 * Content Security Policy directive type.
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy
 */
export type ContentSecurityPolicyDirective = keyof ContentSecurityPolicyRecord;

/**
 * Content Security Policy config type.
 */
export type ContentSecurityPolicyConfig = Partial<ContentSecurityPolicyRecord>;
