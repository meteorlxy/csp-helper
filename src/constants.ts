export const CSP_HEADER_NAME = 'Content-Security-Policy';
