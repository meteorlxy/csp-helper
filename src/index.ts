/* Constants */
export * from './constants';

/* Helpers */
export * from './create-csp-header';
export * from './merge-csp-configs';
export * from './merge-csp-configs-to-set';

/* Presets */
export * from './presets';

/* Types */
export type * from './types';
