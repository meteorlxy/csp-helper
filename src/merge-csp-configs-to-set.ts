import type {
  ContentSecurityPolicyConfig,
  ContentSecurityPolicyDirective,
} from './types';

type ContentSecurityPolicyConfigSet = Partial<
  Record<ContentSecurityPolicyDirective, Set<string>>
>;

/**
 * Helper to merge multiple CSP configs to configs set.
 */
export const mergeCspConfigsToSet = (
  configs: ContentSecurityPolicyConfig[],
): ContentSecurityPolicyConfigSet => {
  const configsSet: ContentSecurityPolicyConfigSet = {};

  for (const config of configs) {
    Object.entries(config).forEach(([key, value]) => {
      const directive = key as ContentSecurityPolicyDirective;
      configsSet[directive] ??= new Set();
      value.split(' ').forEach((v) => configsSet[directive]?.add(v));
    });
  }

  return configsSet;
};
