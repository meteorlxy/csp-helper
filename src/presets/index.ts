export * from './datadog';
export * from './google-ads';
export * from './google-analytics-4';
export * from './google-tag-manager';
export * from './google-universal-analytics';
export * from './hotjar';
