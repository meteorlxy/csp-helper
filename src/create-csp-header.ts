import { CSP_HEADER_NAME } from './constants';
import { mergeCspConfigsToSet } from './merge-csp-configs-to-set';
import type { ContentSecurityPolicyConfig } from './types';

export interface CreateCspHeaderOptions {
  includeHeaderName?: boolean;
  presets?: ContentSecurityPolicyConfig[];
}

export const createCspHeader = (
  config: ContentSecurityPolicyConfig,
  { includeHeaderName = false, presets = [] }: CreateCspHeaderOptions = {},
): string => {
  const mergedConfigsSet = mergeCspConfigsToSet([config, ...presets]);
  const headerValue = Object.entries(mergedConfigsSet)
    .map(([key, value]) => {
      const valueString = [...value].join(' ');
      return valueString ? `${key} ${valueString}` : key;
    })
    .join('; ');
  return includeHeaderName ? `${CSP_HEADER_NAME}: ${headerValue}` : headerValue;
};
