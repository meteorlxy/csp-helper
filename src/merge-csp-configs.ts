import { mergeCspConfigsToSet } from './merge-csp-configs-to-set';
import type { ContentSecurityPolicyConfig } from './types';

/**
 * Helper to merge multiple CSP configs.
 */
export const mergeCspConfigs = (
  configs: ContentSecurityPolicyConfig[],
): ContentSecurityPolicyConfig => {
  const mergedConfigsSet = mergeCspConfigsToSet(configs);

  return Object.fromEntries(
    Object.entries(mergedConfigsSet).map(([key, value]) => [
      key,
      [...value].join(' '),
    ]),
  );
};
