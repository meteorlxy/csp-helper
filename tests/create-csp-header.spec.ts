import { expect, it } from 'vitest';
import { createCspHeader } from '../src/create-csp-header';
import {
  CSP_PRESET_DATADOG_INTAKE_URLS,
  CSP_PRESET_DATADOG_WEB_WORKER,
  CSP_PRESET_HOTJAR,
} from '../src/presets';

it('should create csp header', () => {
  const cspHeader = createCspHeader({
    'base-uri': "'self'",
    'child-src': "'self'",
    'script-src': "'self' https://example.com",
  });
  expect(cspHeader).toEqual(
    "base-uri 'self'; child-src 'self'; script-src 'self' https://example.com",
  );
});

it('should create csp header with header name', () => {
  const cspHeader = createCspHeader(
    {
      'base-uri': "'self'",
      'child-src': "'self'",
      'script-src': "'self' https://example.com",
    },
    {
      includeHeaderName: true,
    },
  );
  expect(cspHeader).toEqual(
    "Content-Security-Policy: base-uri 'self'; child-src 'self'; script-src 'self' https://example.com",
  );
});

it('should create csp header with presets', () => {
  const cspHeader = createCspHeader(
    {
      'base-uri': "'self'",
      'child-src': "'self'",
      'script-src': "'self' https://example.com",
    },
    {
      presets: [
        CSP_PRESET_DATADOG_INTAKE_URLS,
        CSP_PRESET_DATADOG_WEB_WORKER,
        CSP_PRESET_HOTJAR,
      ],
    },
  );
  expect(cspHeader).toMatchSnapshot();
});

it('should create csp header with header name and presets', () => {
  const cspHeader = createCspHeader(
    {
      'base-uri': "'self'",
      'child-src': "'self'",
      'script-src': "'self' https://example.com",
    },
    {
      includeHeaderName: true,
      presets: [
        CSP_PRESET_DATADOG_INTAKE_URLS,
        CSP_PRESET_DATADOG_WEB_WORKER,
        CSP_PRESET_HOTJAR,
      ],
    },
  );
  expect(cspHeader).toMatchSnapshot();
});
