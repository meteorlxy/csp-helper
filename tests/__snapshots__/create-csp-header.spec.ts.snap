// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should create csp header with header name and presets 1`] = `"Content-Security-Policy: base-uri 'self'; child-src 'self'; script-src 'self' https://example.com https://static.hotjar.com https://script.hotjar.com 'unsafe-inline'; connect-src https://*.datadoghq.com https://browser-intake-datadoghq.com https://*.hotjar.com https://*.hotjar.io wss://*.hotjar.com; worker-src blob:; font-src https://script.hotjar.com; img-src https://static.hotjar.com https://script.hotjar.com https://survey-images.hotjar.com; style-src https://static.hotjar.com https://script.hotjar.com 'unsafe-inline'"`;

exports[`should create csp header with presets 1`] = `"base-uri 'self'; child-src 'self'; script-src 'self' https://example.com https://static.hotjar.com https://script.hotjar.com 'unsafe-inline'; connect-src https://*.datadoghq.com https://browser-intake-datadoghq.com https://*.hotjar.com https://*.hotjar.io wss://*.hotjar.com; worker-src blob:; font-src https://script.hotjar.com; img-src https://static.hotjar.com https://script.hotjar.com https://survey-images.hotjar.com; style-src https://static.hotjar.com https://script.hotjar.com 'unsafe-inline'"`;
